be/src/util/bitmap_value.h

```diff
@@ -0,0 +1,1243 @@
+                        }
+                        _bitmap.clear();
+                        break;
+                }
+                break;
+            case BITMAP:
+                switch (_type) {
+                    case EMPTY:
+                        break;
+                    case SINGLE:
+                        if (!rhs._bitmap.contains(_sv)) {
+                            _type = EMPTY;
+                        }
+                        break;
+                    case BITMAP:
+                        _bitmap &= rhs._bitmap;
+                        _convert_to_smaller_type();
+                        break;
+                }
+                break;
+        }
+        return *this;
+    }
+
+    // check if value x is present
+    bool contains(uint64_t x) {
+        switch (_type) {
+            case EMPTY:
+                return false;
+            case SINGLE:
+                return _sv == x;
+            case BITMAP:
+                return _bitmap.contains(x);
+        }
+        return false;
+    }
+
+    // TODO should the return type be uint64_t?
+    int64_t cardinality() const {
+        switch (_type) {
+            case EMPTY:
+                return 0;
+            case SINGLE:
+                return 1;
+            case BITMAP:
+                return _bitmap.cardinality();
+        }
+        return 0;
+    }
+
+    // Return how many bytes are required to serialize this bitmap.
+    // See BitmapTypeCode for the serialized format.
+    size_t getSizeInBytes() {
+        size_t res = 0;
+        switch (_type) {
+            case EMPTY:
+                res = 1;
+                break;
+            case SINGLE:
+                if (_sv <= std::numeric_limits<uint32_t>::max()) {
+                    res = 1 + sizeof(uint32_t);
+                } else {
+                    res = 1 + sizeof(uint64_t);
+                }
+                break;
+            case BITMAP:
+                DCHECK(_bitmap.cardinality() > 1);
+                _bitmap.runOptimize();
+                _bitmap.shrinkToFit();
+                res = _bitmap.getSizeInBytes();
+                break;
+        }
+        return res;
+    }
+
+    // Serialize the bitmap value to dst, which should be large enough.
+    // Client should call `getSizeInBytes` first to get the serialized size.
+    void write(char* dst) {
+        switch (_type) {
+            case EMPTY:
+                *dst = BitmapTypeCode::EMPTY;
+                break;
+            case SINGLE:
+                if (_sv <= std::numeric_limits<uint32_t>::max()) {
+                    *(dst++) = BitmapTypeCode::SINGLE32;
+                    encode_fixed32_le(reinterpret_cast<uint8_t*>(dst), static_cast<uint32_t>(_sv));
+                } else {
+                    *(dst++) = BitmapTypeCode::SINGLE64;
+                    encode_fixed64_le(reinterpret_cast<uint8_t*>(dst), _sv);
+                }
+                break;
+            case BITMAP:
```

Comment:
better add comment that code is written in _bitmap.write.
By the way, the code is written in _bitmap.write and the code is read in BitmapValue. There is some coupling this way. Is there another way to do this?

Reply:
For BITMAP32 and BITMAP64, the type code is both read and write in Roaring64Map. For EMPTY/SINGLE32/SINGLE64, it can't be implemented in Roaring64Map because the purpose of single value optimization is to bypass the RoaringBitmap logic. I think the coupling is OK because Roaring64Map is indeed implementation details of BitmapValue, all clients only see BitmapValue and should never operate on Roaring64Map directly. Maybe I should put Roaring64Map into `detail` namespace to enforce that?

Reply:
OK, I've moved Roaring64Map and related class into `doris::detail` namespace